### PR TITLE
[hail] make Hail PyPI deployment robust

### PR DESCRIPTION
**Pull request by danking**
Changes
---
 - introduce consistent naming of the three types of hail versions (see
   build.gradle comment)

 - add `hail_pip_version` to `_generated_version_info.py` which is used by
   `setup.py`

 - add `hail/python/deploy.sh` which should not be called directly but handles
   generated PYPI compatible packages and uploading to PYPI

 - `generate-build-info.sh` is now the *only* authoritative source on hail
   versioning (it generates *exactly* two files, one for JARs and one for the
   python library)

 - fix: actually activate the anonymous conda env in `hail-ci-build.sh`'s 
   pip test

Janitorial
---
 - ignore python/hail.egg-info

 - sort lines in hail/.gitignore

 - add `hail/python/hail/docs/__init__.py` so that we can use relative imports
   in `conf.py`

 - remove commented out code in makeDocs.sh

 - makeDocs.sh now has `set -x`

Notes
---

You can test deployment locally with:

```
HAIL_TWINE_CREDS_FOLDER=/path/to/twine/creds ./gradlew deploy -D hail.pip-version-suffix=.devN
```
where `.devN` is some currently unused dev version suffix. You can check these by running `python hail/list_pypi_versions.py hail`. These dev versions are not visible to end users.

### CODE REVIEW

File: hail/python/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+     | sed 's/hail (//' \
+     | sed 's/).*//')
+current=$(
+    grep 'hail_pip_version = ' hail/_generated_version_info.py \
+     | sed -E "s/.*hail_pip_version = \"(.*)\"/\1/")
+
+if [[ "${published}" != "${current}" ]]
+then
+    echo deploying ${current}, was ${published}
+    cp ../build/libs/hail-all-spark.jar hail/
```
Comment (tpoterba) [inline hail/python/deploy.sh]: can we clean up this jar below? Or possibly copy the whole python directory to build/tmp first?

seems wrong to leave big artifacts hanging around with no cleanup rules.

File: hail/build.gradle [hunk 2]
```diff
@@ -31,7 +31,13 @@ buildScan {
     licenseAgree = 'yes'
 }
 
-String hailVersion = "0.2"
+// Hail has three notions of version:
+//  - hail short version: MAJOR.MINOR
+//  - hail pip version: MAJOR.MINOR.PATCH
+//  - hail version: MAJOR.MINOR.PATCH-GIT_SHA, calculated by generateBuildInfo
```
Comment (tpoterba) [inline hail/build.gradle]: nice. including patch is a great idea here.

File: hail/python/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+cd $(CDPATH= cd -- "$(dirname -- "$0")" && pwd)
+python3=${HAIL_PYTHON3:-python3}
+
+published=$(
+    pip --no-cache-dir search hail \
+     | grep '^hail ' \
+     | sed 's/hail (//' \
+     | sed 's/).*//')
+current=$(
```
Comment (tpoterba) [inline hail/python/deploy.sh]: I don't love this. We have the version defined in build.gradle - let's just pass it as an argument into this script, or set it as an environment flag.

File: hail/python/hail/docs/conf.py [hunk 4]
```diff
@@ -20,6 +20,10 @@
 import os
 import sys
 
+# sphinx essentially evals this file, so we cannot rely on package structure
```
Comment (tpoterba) [inline hail/python/hail/docs/conf.py]: this seems hacky. What is wrong with environment variables?

I think we have the same dependency structure using env variables or the generated version info file.

File: hail/build.gradle [hunk 5]
```diff
@@ -31,7 +31,13 @@ buildScan {
...
+String hailShortVersion = "0.2"
+// By default pip hides versions ending in ".devN". Use these for testing.
+String hailPipVersion = hailShortVersion + ".4" + System.getProperty("hail.pip-version-suffix", "")
```
Comment (tpoterba) [inline hail/build.gradle]: I don't like calling it "pip version". This is the full Major.Minor.Patch version of Hail, whether we deploy it to pip or conda or wherever.

File: hail/python/deploy.sh (hunk 1 above)
Comment (danking) [reply hail/python/deploy.sh]: I'm really not a fan of copying all the python stuff around. I want to eventually understand the docs build enough to get rid of it. I'll add a trap to cleanup the jar.

File: hail/python/deploy.sh (hunk 3 above)
Comment (danking) [reply hail/python/deploy.sh]: Hmm. I see what you mean about this, but I'm also not a fan of having two sources of truth. If something gets out of sync we could deploy a version of hail whose `_generated_version_info.py` disagrees with PyPI's version.

Comment (danking) [reply hail/python/deploy.sh]: I could pass it in and verify it matches this expression exactly?

File: hail/build.gradle (hunk 5 above)
Comment (danking) [reply hail/build.gradle]: I wanted to call it `hailFullVersion` or `hailLongVersion` but both of those seem like they should refer to the version with the git sha. I'm at a loss for an appropriate name for this.

**Comment by danking** [comment]
Two of your comments deal with my elimination of redundant sources of information, so I'll address both here.

I'm basically thinking of us when we have to debug this system. It's confusing if there's two sources of truth or if we're manually calling `deploy.sh` to isolate issues with that from issues with gradle, I don't want to have two separate knobs to spin. If they accidentally get out of sync that's gonna be double confusing (imagine a PyPI version that disagrees with hail's internal version).

How about we put these two versions into two single-line, plain text files and have _generated... load the files to initialize the variables?

**Comment by tpoterba** [comment]
oops, meant to respond that I'm totally fine with including the version in a plaintext file.

File: hail/build.gradle (hunk 5 above)
Comment (tpoterba) [reply hail/build.gradle]: patch version?

**Comment by danking** [comment]
I also added a trap to cleanup the jar

**Pull request merged**